### PR TITLE
Create JEI-like Advanced Search

### PR DESCRIPTION
**Pull request by slprime**
## Advanced search:
`@` - search by mod (`@minecraft`)
`#` - search by tooltip (`#UV-Tier`)
`&` - search by id: (`&95:15` `extratrees:door`)
`$` - search by ore dictionary (`$ore`)
`%` - search by subsets (`%block`)
`|` - logical or. multi-item search (`wrench|hammer`)
`-` - logical not. exclude items that match the following expression (`-@minecraft`)
`"` - multiple words wrapped in quotes are treated as one condition rather than several (`"White Stained Glass"`)

### Search Mode (apply on search parts, not full string):
`Plain` - simple string
`Extended` - allow only `?`, `*` rules
`Regex` - allow full regex

### In Extended Search Mode:
`?` - any character (`@min?craft`)
`*` - any few characters (`@m*ft` = `@minecraft`)
`r/.../` - standard java regex (`@r/m\w{6}ft/` = `@minecraft`)

## More colors:
image

## Settings:
image

## P.S.
1. issue https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/15934
1. Similar PR (https://github.com/GTNewHorizons/NotEnoughItems/pull/480)
1. WE NEED someone who understands how `NotEnoughCharacters` works and whether the behavior is preserved. and how much this mod is needed by a new search. I don't know Chinese and can't check.

### P.S.2
1. If this PR is approved i started work with AE2 terminals and holo inventory to update search rules there
2. If exists ability to get GT ItemStack Tier, add search prefix for it will be excellent

### CODE REVIEW

**Comment by slprime** [comment]
image

move `Focus Search Widget on Open` config option to Search Widget scope

**Comment by OvermindDL1** [comment]
> move `Focus Search Widget on Open` config option to Search Widget scope, and rename

Just to make sure, the "Search Widget Scope" is a new settings category that does not exist as-is in `master` right now, correct?

Is there anything in this PR I could assist with in the next few days?  Any open sub-issues?

File: src/main/java/codechicken/nei/ClientHandler.java [hunk 1]
```diff
@@ -149,6 +149,7 @@ public static void preInit() {
         loadSerialHandlers();
         loadHeightHackHandlers();
         loadHiddenHandlers();
+        loadEnableAutoFocus();
```
Comment (OvermindDL1) [inline src/main/java/codechicken/nei/ClientHandler.java:152]: Oh bugger, I completely forgot that, thanks!  Should I PR it to master or just leave it for this PR?

**Comment by slprime** [comment]
> Just to make sure, the "Search Widget Scope" is a new settings category that does not exist as-is in `master` right now, correct?

yes, this scope added in this PR

**Comment by vfyjxf** [comment]
> 3. WE NEED someone who understands how `NotEnoughCharacters` works and whether the behavior is preserved. and how much this mod is needed by a new search. I don't know Chinese and can't check.
> 

NotEnoughCharacters just register a SearchField.ISearchProvider to support jei search style and extra search support.

**Comment by vfyjxf** [comment]
And are you using the same tree search as jei? If you ported it, then nech may need additional changes to be more compatible with nei.

**Comment by slprime** [comment]
> And are you using the same tree search as jei? If you ported it, then nech may need additional changes to be more compatible with nei.

No, I don't use the JEI tree. I've changed the search field parser. In the current version, prefixes only worked at the beginning of the search text. Now, it's split into tokens, and an ItemFilter is generated based on them. The only change I made was to deprecate ISearchProvider and add ISearchParserProvider, so that other mods wouldn't break with the new search.

**Comment by slprime** [comment]
In `NotEnoughCharacters` exists `@`, `$`, `&` which are already added the new search

**Comment by vfyjxf** [comment]
> In NotEnoughCharacters exists @, $, & prefixes which are already added the new search

yes,and chinese support also dependends on ISearchProvider

**Comment by slprime** [comment]
> > In NotEnoughCharacters exists @, $, & prefixes which are already added the new search
> 
> yes,and chinese support also dependends on ISearchProvider

If I understand the `net.minecraft.nechar.Nechar SearchFilter` class correctly, it is not needed in the new search.

**Comment by vfyjxf** [comment]
> If I understand the net.minecraft.nechar.Nechar SearchFilter class correctly, it is not needed in the new search.
 
No, matchesKeywords method provide support of chinese pinyin search

**Comment by slprime** [comment]
> > If I understand the net.minecraft.nechar.Nechar SearchFilter class correctly, it is not needed in the new search.
> 
> No, matchesKeywords method provide support of chinese pinyin search

Then you can replace the default oreDict filter

`API.addSearchProvider(new SearchParserProvider( '$', "oreDict", EnumChatFormatting.AQUA, (pattern) -> new OreDictionaryFilter(pattern)));`

or added to it subfilter

`SearchField.searchParserProviders.get('$').addSubFilter((pattern) -> new OreDictionaryFilter(pattern))`

**Comment by slprime** [comment]
> @slprime also,can you provide an api that allows a certain item filter to be enabled only if the game is in a certain language, like the input method api for rei.

yes, I'll add it after work

**Comment by slprime** [comment]
@vfyjxf, you need api which apply custom ISearchParserProvider only if the game is in a certain language, else use default ISearchParserProvider? correct?

**Comment by vfyjxf** [comment]
> @vfyjxf, you need api which apply custom ISearchParserProvider only if the game is in a certain language, else use default ISearchParserProvider? correct?

YES, it helps nech not to affect non-Chinese users.

**Comment by ah-OOG-ah** [comment]
1. How much of a change is this from the default search?
2. Is this behind a config, or on by default?

**Comment by chochem** [comment]
Just to be clear: do I understand correctly that this changes the connection between terms from OR to AND? Then one gets OR instead with `|`?

I am not even saying I hate that, but that was one of the reasons NECh was rejected.

**Comment by chochem** [comment]
there seems to be another issue, in that it no longer searches the internal item names which results in very different results. This is certainly going to be controversial.

**Comment by chochem** [comment]
here is another issue from testing: I cant search for blood magic items (assuming I am a player who doesnt know the secret password is `@awwayof`). something like `#"blood magic"` does not work. I guess the tooltip search is somewhat broken or am I doing something wrong?

**Comment by slprime** [comment]
## Prefix Schema:
image

## Item Subset (`%`). In previous version you use `@` to search by this list. Now `@` usage for modname (jei style)
image

## If you want restore NEI search, you may change config like in image below
image

## But have some breaking changes:
1. to search by subsets you need to use `%`, not `@`
2. `|` cannot be disabled

**Comment by slprime** [comment]
> here is another issue from testing: I cant search for blood magic items (assuming I am a player who doesnt know the secret password is `@awwayof`). something like `#"blood magic"` does not work. I guess the tooltip search is somewhat broken or am I doing something wrong?

fix modname. please use `@"blood magic"`

image

**Comment by slprime** [comment]
> Edit: I have been told that this is what you meant by id in your description. Which is well definetely not correctly named then, But adding `&` does indeed work.

naming from JEI/REI

**Comment by slprime** [comment]
> Just to be clear: do I understand correctly that this changes the connection between terms from OR to AND? Then one gets OR instead with `|`?
> 
> I am not even saying I hate that, but that was one of the reasons NECh was rejected.

image

NEI Search Style (old style)

Item Name/Tooltip/Mod Name include: 
rune`<space>`of`<space>`super`<space>`|`<space>`green`<space>`backpack

JEI/REI Search Style (new style). `|` split search to 2 search rules. like OreDict Card

Item Name include: run`<or>`of`<or>`super
or Item Name include: green `<or>` backpack

p.s. sorry for my english )

**Comment by slprime** [comment]
> 1. How much of a change is this from the default search?
> 2. Is this behind a config, or on by default?

1. previous search looked for in full tooltip. now only in display name (if you don't use prefixes)
2. https://github.com/GTNewHorizons/NotEnoughItems/pull/485#issuecomment-2176436314

**Comment by evgengoldwar** [comment]
Fantastic change. The new versions of Jei have done the same thing, and it's much more convenient than what Nei has now. Looking forward to seeing this added to the game.

**Comment by slprime** [comment]
leaving two search options is a bad idea. the code will become much more complex. the new search can work like the old search by changing some settings (in the comment above), not 100% but very close. if you search by name/tooltip/mod name/id or use `@` to search for a mod, that will enough

**Comment by slprime** [comment]
image

**Comment by slprime** [comment]
How search works (3:50-9:30) https://youtu.be/c77g2imldpk?si=m37tmz3pN_CUAF_n&t=229

**Comment by vfyjxf** [comment]
@slprime I would like to ask if this situation is a bug? IConfigureNEI.loadConfig is later than SearchTokenParser.getProviders, usually mod authors choose to register ISearchParserProvider via NEIConfig, but SearchTokenParser has kept a cache before that, which causes ISearchParserProvider registered by other mods to be unavailable unless the player switches languages and clears the cache.

**Comment by vfyjxf** [comment]
And how do I override the original ISearchParserProvider, obviously nech needs to override the prefix-less and tooltip ISearchParserProvider to support its additional search functionality, mixin is probably not a good idea.

**Comment by slprime** [comment]
> SearchParserProvider

> And how do I override the original ISearchParserProvider, obviously nech needs to override the prefix-less and tooltip ISearchParserProvider to support its additional search functionality, mixin is probably not a good idea.

you need registry API.addSearchProvider like in this example or not suitable?
https://github.com/GTNewHorizons/NotEnoughItems/pull/485#issuecomment-2164458039

**Comment by slprime** [comment]
```

```

> > > SearchParserProvider
> > 
> > 
> > > And how do I override the original ISearchParserProvider, obviously nech needs to override the prefix-less and tooltip ISearchParserProvider to support its additional search functionality, mixin is probably not a good idea.
> > 
> > 
> > you need registry API.addSearchProvider like in this example or not suitable? [#485 (comment)](https://github.com/GTNewHorizons/NotEnoughItems/pull/485#issuecomment-2164458039)
> 
> but [this](https://github.com/GTNewHorizons/NotEnoughItems/blob/4f98ab6c646967e32cdce04c8fe8c4cdfe35cbf5/src/main/java/codechicken/nei/SearchTokenParser.java#L102) reject my provider i think

you provider should be added after the default. this condition searches for the first from the end

**Review by Ethryan — APPROVED** [review]
Seems to be working as intended.
And I havn't run into any issues during my testing.

**Pull request merged**